src/com/owncloud/android/ui/preview/PreviewUtils.java

```diff
@@ -0,0 +1,149 @@
+import android.util.Base64;
+
+import com.google.android.exoplayer2.upstream.DataSource;
+import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
+import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
+import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory;
+import com.google.android.exoplayer2.upstream.HttpDataSource;
+import com.owncloud.android.MainApp;
+import com.owncloud.android.datamodel.OCFile;
+import com.owncloud.android.lib.common.OwnCloudBasicCredentials;
+import com.owncloud.android.lib.common.OwnCloudCredentials;
+import com.owncloud.android.lib.common.OwnCloudSamlSsoCredentials;
+import com.owncloud.android.lib.common.accounts.AccountUtils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+
+/**
+ * Created by davidgonzalez on 22/2/17.
+ *
+ * An utils provider for building data sources used in video preview
+ */
+
+public class PreviewUtils {
+
+    private static final DefaultBandwidthMeter BANDWIDTH_METER = new DefaultBandwidthMeter();
+
+    /**
+     * Returns a new DataSource factory.
+     *
+     * @param useBandwidthMeter Whether to set {@link #BANDWIDTH_METER} as a listener to the new
+     *                          DataSource factory.
+     * @return A new DataSource factory.
+     */
+    protected static DataSource.Factory buildDataSourceFactory(boolean useBandwidthMeter,
+                                                               Context context, OCFile file,
+                                                               Account account) {
+        return buildDataSourceFactory(useBandwidthMeter ? BANDWIDTH_METER : null, context, file, account);
+    }
+
+    protected static DataSource.Factory buildDataSourceFactory(DefaultBandwidthMeter bandwidthMeter,
+                                                               Context context, OCFile file, Account account) {
+        return new DefaultDataSourceFactory(context, bandwidthMeter,
+                buildHttpDataSourceFactory(bandwidthMeter, file, account));
+    }
+
+    /**
+     * Returns a new HttpDataSource factory.
+     *
+     * @param bandwidthMeter Whether to set {@link #BANDWIDTH_METER} as a listener to the new
+     *     DataSource factory.
+     * @return A new HttpDataSource factory.
+     */
+    protected static HttpDataSource.Factory buildHttpDataSourceFactory(DefaultBandwidthMeter bandwidthMeter,
+                                                                       OCFile file, Account account) {
+
+        if (file.isDown()) {
+
+            return new DefaultHttpDataSourceFactory(MainApp.getUserAgent(), bandwidthMeter);
+
+        } else {
+
+            try {
+
+                //Get account credentials asynchronously
+                final GetCredentialsTask task = new GetCredentialsTask();
+                task.execute(account);
+
+                OwnCloudCredentials credentials = task.get();
```

Comment:
This will not work fine.

You run a task as an ```AsyncTask``` because you want it occurs in another thread, because the task is too long and don't want the current thread (probably the main thread) uses all that time in this.

But then you call ```task.get()```, and that **waits** in the current thread until the task is finished.

So, in the end, this works just exactly as if you do not use an ```AsyncTask```, but with more code. Not a good deal.